document-portal/document-portal-fuse.c

```diff
@@ -3079,11 +3145,58 @@ xdp_fuse_mainloop (gpointer data)
       g_file_set_contents (status, s->str, -1, &error);
       g_assert_no_error (error);
     }
+}
 
-  fuse_session_remove_chan (main_ch);
-  fuse_session_destroy (session);
-  fuse_unmount (mount_path, main_ch);
-  main_ch = NULL;
+static gpointer
+xdp_fuse_thread (gpointer data)
+{
+  g_autoptr(GMutexLocker) locker = NULL;
+  g_autoptr (GMutexLocker) session_locker = NULL;
+  const char *path;
+  struct fuse_session *se;
+  XdpFuseThreadData *thread_data = data;
+  struct fuse_loop_config loop_config = {0};
+
+  locker = g_mutex_locker_new (&thread_data->lock);
+  fuse_pthread = pthread_self ();
+
+  se = fuse_session_new (thread_data->args, &xdp_fuse_oper,
+                         sizeof (xdp_fuse_oper), NULL);
+  if (se == NULL)
+    {
+      g_set_error (&thread_data->error, XDG_DESKTOP_PORTAL_ERROR,
+                   XDG_DESKTOP_PORTAL_ERROR_FAILED,
+                   "Can't create fuse session");
+      g_cond_signal (&thread_data->cond);
+      return NULL;
+    }
+
+  path = xdp_fuse_get_mountpoint ();
+  if (fuse_session_mount (se, path) != 0)
+    {
+      fuse_session_destroy (se);
+      g_set_error (&thread_data->error, XDG_DESKTOP_PORTAL_ERROR,
+                   XDG_DESKTOP_PORTAL_ERROR_FAILED,
+                   "Can't mount path %s", path);
+      g_cond_signal (&thread_data->cond);
+      return NULL;
+    }
+
+  loop_config = thread_data->loop_config;
+  session = se;
+
+  g_cond_signal (&thread_data->cond);
```

Comment:
My earlier suggestion was to put this signal() call right after the g_mutex_locker_new() call, and I think that's still a good idea. The `xdp_fuse_init` function cannot continue until we release the mutex, so it doesn't matter where in the critical section we make the call.

And by doing it once at the beginning before any branches, we don't need to worry that one of the error paths will forget to signal.